Feat(eos_cli_config_gen): Add bandwidth_guaranteed_percent for QoS Profiles.

pibjndata opened the pull request:
Adding possibility to configure new missing "bandwidth guaranteed percent <rate>" command.
The issue appeared on device 7010T-48 with software: 4.25.5.1M .
New port_profile definition must have bandwidth_guaranteed_percent variable instead of bandwidth_percent.

## Change Summary

- Add bandwidth_guaranteed_percent option to QoS data format configuration

- Add bandwidth_guaranteed_percent to upload documentation correctly 

## Related Issue(s)

Fixes #1304

## Component(s) name

`arista.avd.eos_cli_config_gen`

## Proposed changes

create qos profile and run on device 7010T-48 with software: 4.25.5.1M 
# QOS profiles
qos_profiles:
  QOS-PROFILE:
    trust: dscp
    tx_queues:
      1:
        bandwidth_guaranteed_percent: 10
      2:
        bandwidth_guaranteed_percent: 20

## How to test
Run fabric playbook with new variables

## Checklist

- New command appears if new variable defined

### User Checklist
- N/A

### Repository Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

pibjndata committed:
Feat(eos_cli_config_gen): Adding new command syntax for QoS Profiles.

Adding possibility to configure new missing "bandwidth guaranteed percent <rate>" command.
The issue appeared on device 7010T-48 with software: 4.25.5.1M .
New port_profile definition must have bandwidth_guaranteed_percent variable instead of bandwidth_percent.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/qos-profiles.j2:
```diff
@@ -24,7 +24,13 @@ QOS Profile: **{{ profile }}**
 {%             for tx_queue in qos_profiles[profile].tx_queues | arista.avd.natural_sort %}
 {%                 set queue = qos_profiles[profile].tx_queues[tx_queue] | default('-') %}
-{%                 set bw_percent = qos_profiles[profile].tx_queues[tx_queue].bandwidth_percent | default('-') %}
+{%                 if qos_profiles[profile].tx_queues[tx_queue].bandwidth_percent is arista.avd.defined %}
```
I suggest to refactor to use the `arista.avd.default` filter's ability to test multiple values.
Example:
```jinja
{% set bw_percent = qos_profiles[profile].tx_queues[tx_queue].bandwidth_perfect | arista.avd.default(
                    qos_profiles[profile].tx_queues[tx_queue].bandwidth_guaranteed_percent,
                    '-') %}
```

pibjndata replied:
Good idea!

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/qos-profiles.j2:
````diff
@@ -36,4 +42,4 @@ QOS Profile: **{{ profile }}**
 {% include 'eos/qos-profiles.j2' %}
 ```
-{% endif %}
+{% endif %}
````
Missing final newline.

pibjndata replied:
done

ClausHolbechArista replied:
still missing...

ClausHolbechArista reviewed (changes requested):
Thank you for your contribution. Please adjust the code as described and then we will need to add CI artifacts for this afterwards.

pibjndata commented:
@ClausHolbechArista 
Hey,
just a request for second review ;)

ClausHolbechArista commented:
> @ClausHolbechArista 
> Hey,
> just a request for second review ;) 

Sure I will take a look over the weekend.

ClausHolbechArista reviewed (changes requested):
I did some small style changes to your files, to get the documentation template completed.
Also a single line had been deleted from the config template, but I have added it again.
Please make sure to test your work locally before submitting. Ex running a small inventory on your laptop using the new keys.
**Make sure you pull from your fork again before continuing.**

Next is to add molecule coverage of the new key. Molecule is what we use to run various test scenarios in the CI pipeline. It is small ansible plays with controlled input and expected output.
Check the file `ansible_collections/arista/avd/molecule/eos_cli_config_gen/inventory/host_vars/qos.yml`, and add another qos_profile with your new key. You don't need to reference the profile on an interface, since you did not change any logic.
To run the tests locally, first install all development requirements `pip3 install -r development/dev-requirements.txt` and then run molecule:
```cli
cd ansible_collections/arista/avd/molecule
make refresh-facts
```
It takes a long time to run - depending on your resources.
When the molecule run is done, you can check what files changed, and if the changes are as you expect. When you are satisfied, you commit your changes including molecule input/output, and then the same run will happen in the CI pipeline on GitHub. If they match, the test will succeed. Then we as reviewers will verify if your input and output is as you have documented and can merge the PR.

tgodaA commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/qos-profiles.j2:
```diff
@@ -16,6 +16,8 @@ qos profile {{ profile }}
    tx-queue {{ tx_queue }}
 {%         if qos_profiles[profile].tx_queues[tx_queue].bandwidth_percent is arista.avd.defined %}
       bandwidth percent {{ qos_profiles[profile].tx_queues[tx_queue].bandwidth_percent }}
+{%         elif qos_profiles[profile].tx_queues[tx_queue].bandwidth_guaranteed_percent is arista.avd.defined %}
```
Bandwidth guaranteed should be with in qos profile but not inside tx-queue:
```
CL(config-qos-profile-test1)#bandwidth ?
  guaranteed  Set guaranteed bandwidth

```

FYI, with in tx-queue the bandwidth options:
```
CL(config-qos-profile-test1-txq-0)#bandwidth ?
  percent  Set proportion of the total bandwidth when scheduling is round-robin
```

ClausHolbechArista replied:
this is hardware dependent, which is what the customer hit in this case. So we do need the guaranteed percent under the tx queue on some platforms.

tgodaA replied:
Ok. In that case would it be ok to add the same knob under qos profile as well along with adding it to the tx-queue.

ClausHolbechArista replied:
I don't think we should. I have tried on 7010, 7050X3 and 7020SR and none of them accept the command directly on the profile. I think it is a bug that it is available at that level, since the help text suggests that it should be under the queue:
```
cdv471(config-s-test-qos-profile-test)#?
  abort                  Abandon all changes
  bandwidth              Configure guaranteed/round-robin minimum bw for this queue
...
```

pibjndata committed:
adding changes after make refresh-facts

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/qos-profiles.j2:
```diff
@@ -16,9 +16,11 @@ qos profile {{ profile }}
 {%         endif %}
 {%     endfor %}
-{% endfor %}
+{% endfor %}
```
missing newline at the end of the file

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/documentation/qos-profiles.j2:
```diff
@@ -12,20 +12,22 @@ QOS Profile: **{{ profile }}**
-{%                 set bw_percent = qos_profiles[profile].tx_queues[tx_queue].bandwidth_percent | default('-') %}
-{%                 set priority = qos_profiles[profile].tx_queues[tx_queue].priority | default('-') %}
+{%                  set queue = qos_profiles[profile].tx_queues[tx_queue] | arista.avd.default('-') %}
+{%                  set bw_percent = qos_profiles[profile].tx_queues[tx_queue].bandwidth_perfect | arista.avd.default(
```
typo, `perfect` instead of `percent`. This causes the unintended changes in the molecule output.

ClausHolbechArista reviewed (changes requested):
There is some extra file added (by mistake I'm sure).
`ansible_collections/arista/avd/molecule/=3.2.0,`

pibjndata committed:
Feat(eos_cli_config_gen): Adding new command syntax for QoS Profiles.
Adding possibility to configure new missing "bandwidth guaranteed percent <rate>" command.
The issue appeared on device 7010T-48 with software: 4.25.5.1M .
New port_profile definition must have bandwidth_guaranteed_percent variable instead of bandwidth_percent.
Last changes:
- removed typo from perfect to percent
- added empty lines at the end of files
- added molecule documentation files
- removed added by mistake =3.2.0, file

ClausHolbechArista reviewed (approved):
LGTM

carlbuchmann committed:
update CI artifacts

carlbuchmann reviewed (approved):
LGTM!
